test/features/counter/Counter.spec.tsx

```diff
@@ -0,0 +1,155 @@
+  it('should should display count', () => {
+    const { p } = setup();
+    expect(p.text()).toMatch(/^1$/);
+  });
+
+  it('should first button should call increment', () => {
+    const { buttons, actions } = setup();
+    const incrementSpy = spy(actions, 'increment');
+
+    buttons.at(0).simulate('click');
+    expect(incrementSpy.called).toBe(true);
+    incrementSpy.restore();
+  });
+
+  it('should match exact snapshot', () => {
+    const { store } = setup();
+    const tree = renderer
+      .create(
+        <Provider store={store}>
+          <Router>
+            <Counter />
+          </Router>
+        </Provider>
+      )
+      .toJSON();
+
+    expect(tree).toMatchSnapshot();
+  });
+
+  it('should second button should call decrement', () => {
+    const { buttons, actions } = setup();
+    const decrementSyp = spy(actions, 'decrement');
+    buttons.at(1).simulate('click');
+    expect(decrementSyp.called).toBe(true);
+    decrementSyp.restore();
+  });
+
+  it('should third button should call incrementIfOdd', () => {
+    const { buttons, actions } = setup();
+    const incrementIfOdd = spy(actions, 'incrementIfOdd');
+    buttons.at(2).simulate('click');
+    expect(incrementIfOdd.called).toBe(true);
+    incrementIfOdd.restore();
+  });
+
+  it('should fourth button should call incrementAsync', () => {
+    const { buttons, actions } = setup();
+    const incrementAsync = spy(actions, 'incrementAsync');
+    buttons.at(3).simulate('click');
+    expect(incrementAsync.called).toBe(true);
+    incrementAsync.restore();
+  });
+
+  it('should display updated count after increment button click', () => {
+    const { buttons, p } = setup();
+    buttons.at(0).simulate('click');
+    expect(p.text()).toMatch(/^2$/);
+  });
+
+  it('should display updated count after decrement button click', () => {
+    const { buttons, p } = setup();
+    buttons.at(1).simulate('click');
+    expect(p.text()).toMatch(/^0$/);
+  });
+
+  it('shouldnt change if even and if odd button clicked', () => {
+    const { buttons, p } = setup({ counter: { value: 2 } });
+    buttons.at(2).simulate('click');
+    expect(p.text()).toMatch(/^2$/);
+  });
+
+  it('should change if odd and if odd button clicked', () => {
+    const { buttons, p } = setup({ counter: { value: 1 } });
+    buttons.at(2).simulate('click');
+    expect(p.text()).toMatch(/^2$/);
+  });
+});
+
+describe('Test counter actions', () => {
+  const actions = counterActions;
+  it('should increment should create increment action', () => {
+    expect(actions.increment()).toMatchSnapshot();
+  });
+
+  it('should decrement should create decrement action', () => {
+    expect(actions.decrement()).toMatchSnapshot();
+  });
+
+  it('should incrementIfOdd should create incrementIfOdd action', () => {
+    expect(actions.decrement()).toMatchSnapshot();
+  });
+
+  it('should not call incrementAsync before timer', () => {
+    const fn = actions.incrementAsync(1000) as Function;
```

Comment:
Honestly, I hate this, but I had to cast to something because by default it is well aware of the fact it is an AppThunk and expects 3 parameters for `fn()` call. I don't really know what should I do there, knowing about AppThunk and passing a dummy `getSelect` everywhere with a proper `RootState` looks like a lot of unnecessary work, as I only want to test a mocked dispatch:/

Reply:
Feels like this is maybe too specific somehow?

Reply:
Honestly, I don't know, there was a test originally to check if the action fired after a timeout, but I think it is totally reasonable to test it to NOT fire faster than we want.
But I could check it on component too, but that means tests on actions alone are totally useless. I am really not sure about that. If you guys vote to remove it I am happy to do so.